src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -231,47 +231,46 @@ internal ItemManager(RecentItemsManager recentItemsManager, IGlobalOptionService
+
+                if (!filteringResult.HasValue)
+                    return null;
 
+                var (selectedItemIndex, selectionHint, uniqueItem) = filteringResult.Value;
+                var useAggressiveDefaultsMatching = session.TextView.Options.GetOptionValue<bool>(AggressiveDefaultsMatchingOptionName);
+                var showCompletionItemFilters = _globalOptions.GetOption(CompletionViewOptions.ShowCompletionItemFilters, document?.Project.Language);
                 var updatedFilters = showCompletionItemFilters
                     ? GetUpdatedFilters(initialListOfItemsToBeIncluded, data.SelectedFilters)
                     : ImmutableArray<CompletionFilterWithState>.Empty;
 
-                var useAggressiveDefaultsMatching = session.TextView.Options.GetOptionValue<bool>(AggressiveDefaultsMatchingOptionName);
-
-                // If this was deletion, then we control the entire behavior of deletion ourselves.
-                if (initialRoslynTriggerKind == CompletionTriggerKind.Deletion)
-                {
-                    return HandleDeletionTrigger(reason, initialListOfItemsToBeIncluded, filterText, updatedFilters, hasSuggestedItemOptions, highlightMatchingPortions, completionHelper);
-                }
+                return new FilteredCompletionModel(
+                    items: GetHighlightedList(initialListOfItemsToBeIncluded, filterText, highlightMatchingPortions, completionHelper),
+                    selectedItemIndex,
+                    updatedFilters,
+                    selectionHint,
+                    centerSelection: true,
+                    uniqueItem);
+            }
+            finally
+            {
+                // Don't call ClearAndFree, which resets the capacity to a default value.
+                initialListOfItemsToBeIncluded.Clear();
+                s_listOfMatchResultPool.Free(initialListOfItemsToBeIncluded);
+            }
 
-                Func<ImmutableArray<(RoslynCompletionItem, PatternMatch?)>, string, ImmutableArray<RoslynCompletionItem>> filterMethod;
+            Func<ImmutableArray<(RoslynCompletionItem, PatternMatch?)>, string, ImmutableArray<RoslynCompletionItem>> GetFilterMethod()
+            {
                 if (completionService == null)
                 {
-                    filterMethod = (itemsWithPatternMatches, text) => CompletionService.FilterItems(completionHelper, itemsWithPatternMatches, text);
+                    return (itemsWithPatternMatches, text) => CompletionService.FilterItems(completionHelper, itemsWithPatternMatches, text);
                 }
                 else
                 {
                     Contract.ThrowIfNull(document);
-                    filterMethod = (itemsWithPatternMatches, text) => completionService.FilterItems(document, itemsWithPatternMatches, text);
+                    return (itemsWithPatternMatches, text) => completionService.FilterItems(document, itemsWithPatternMatches, text);
```

Comment:
i'm so confused by this code.  it looks nearly identical... just one access the static on the type, but one the instance?  this needs explaining.